conbench/entities/distribution.py

```diff
@@ -108,109 +115,122 @@ def get_distribution(benchmark_result, limit):
+    ending with the most recent BenchmarkResult in the given BenchmarkResult's commit
+    ancestry on the default branch.
+
+    We choose a commit from the default branch as the end commit, because we probably
+    don't want "bad" commits early on in a PR to set artificially low z-scores later in
+    the PR.
+
+    For the start commit of the distribution, uses whatever commit_limit was used when
+    writing to the Distribution table.
+
+    Should not raise anything. If we can't find a z-score for some reason, the z_score
+    attribute will be None on that BenchmarkResult.
+    """
+    # Cache Distribution query results by commit/hardware, in case multiple input
+    # benchmark_results (e.g. from the same Run) share those attributes
+    cached_distributions = {}
+    # Also cache respective baseline runs by each Run's ID. Note that two
+    # benchmark_results from the same Run might not have the exact same baseline run,
+    # because one baseline run might be missing the case/context of the other
+    # benchmark_result. In those cases, we might fail to populate the z-score of the
+    # second benchmark result. But that's too rare and small-impact to justify not
+    # caching for performance reasons.
+    cached_baseline_runs = {}
 
     for benchmark_result in benchmark_results:
         benchmark_result.z_score = None
+        if benchmark_result.error:
+            log.debug("benchmark_result has error; setting z_score = None")
+            continue
 
-    first = benchmark_results[0]
-    parent_commit = get_closest_parent(first.run)
-    if not parent_commit:
-        return
-
-    where = [
-        Distribution.commit_id == parent_commit.id,
-        Distribution.hardware_hash == first.run.hardware.hash,
-    ]
-    if len(benchmark_results) == 1:
-        where.extend(
-            [
-                Distribution.case_id == first.case_id,
-                Distribution.context_id == first.context_id,
-            ]
-        )
+        if benchmark_result.run_id in cached_baseline_runs:
+            baseline_run = cached_baseline_runs[benchmark_result.run_id]
+        else:
+            baseline_run = benchmark_result.run.get_baseline_run(
+                on_default_branch=True,
+                case_id=benchmark_result.case_id,
+                context_id=benchmark_result.context_id,
+            )
+            cached_baseline_runs[benchmark_result.run_id] = baseline_run
 
-    cols = [
-        Distribution.case_id,
-        Distribution.context_id,
-        Distribution.mean_mean,
-        Distribution.mean_sd,
-    ]
+        if not baseline_run:
+            log.debug("No baseline run; setting benchmark_result.z_score = None")
+            continue
 
-    distributions = Session.query(*cols).filter(*where).all()
-    lookup = {f"{d.case_id}-{d.context_id}": d for d in distributions}
+        commit = baseline_run.commit_id
+        hardware = baseline_run.hardware.hash
```

Comment:
This might be only an issue while reviewing this code (given the large scope of changes), but I wonder if we should name these `baseline_commit` and `baseline_hardware` to make sure it's clear that that's what they are later on.

When reviewing this, at first, I was worried that the line below in 211 was actually using the distribution starting from the contender and not the baseline and was super confused until I came back to find this

Reply:
sure!